feat: Find current scope function.

tagyieh opened the pull request:
Implemented function to determine what node owns the scope.

Closes #30

tagyieh committed:
feat: Find current scope function.

Implemented function to determine what node owns the scope.

Tobiky commented on src/malls/mal_lsp.py:
Please move the newly defined functions into a new file under module "ts" or "treesitter", since these functions  work independently from the language server as well as any instance of the server (so they should have been static/class methods instead, if attached to `MALLSPServer`)

Tobiky commented on src/malls/mal_lsp.py:
```diff
@@ -224,3 +228,85 @@ def m___set_trace(self, **params: dict | None) -> None:
+                        owner = root_node
+                    break
+                case _:
+                    pass
```
Are all breaks as intended? Some appear outside of the if-statements but seem to serve the same purpose so I wish to double check.

Furthermore, you can simply the statement by merging cases. In the scenario that all breaks should be inside the if-statment the match statement can look like this:

```python
            match cursor.node.type:
                case "category_declaration" | "asset_declaration" | "associations_declaration" as node_type:
                    # we have to check if the point is before or after the '{'
                    # we can use query to find this
                    owner = cursor.node

                    if (self._query_and_compare_scope_pos(node_type, cursor, point)):
                        owner = root_node
                        break
                case _:
                    pass
```

tagyieh replied:
I believe the breaks do indeed have to be like this:
- `category_declaration`: if the position happens before the `{`, then we can break right away and assume that the scope is the root node. Otherwise, we need to keep iterating the loop, to find the innermost parent
- `asset_declaration`: if the condition is true, then we know that the parent of this node, a `category_declaration` is the owner of the scope, hence the `owner = owner.parent`. Regardless, an asset is the innermost parent that can own a scope, so we break right away (after finding if the position is before or after `{`)
- `associations_declaration`: is a mix of the two previous. Like the category, if the position is before `{` then the owner is the root node. Like the asset, there is no child of an association that can own a scope, hence the break outside the if.

I think that for these reasons they cannot be merged into smaller conditions

Tobiky commented on tests/features/test_find_current_scope.py:
```diff
@@ -0,0 +1,96 @@
+from tree_sitter import Language, Parser, QueryCursor
+import tree_sitter_mal as ts_mal
+
+FILE = open("./tests/mal_fixtures/test_find_current_scope.mal","rb")
```
Please use the automatic fixture loading and definition by putting file fixtures instead `tests/fixtures`. If you wish to sub-folder them you can do so where `tests/fixtures/mal/test_find_current_scope.mal` will become available as `mal_test_find_current_scope` (i.e. replacing `/` with `_ `). If you wish to change how this works you can do so in `conftest.py` in the root of `tests`.

Also avoid the pytest naming convention in the fixtures since it may confuse pytest, efforts in finding the actual tests, and increases the name length.

Tobiky commented on tests/features/test_find_current_scope.py:
```diff
@@ -0,0 +1,96 @@
+        super().__init__(io.BytesIO, io.BytesIO)
+
+    def find_current_scope(self, cursor, point):
+        return self._find_current_scope(cursor, point)
```
Why not use directly or at least extend the one that exists in `util`?

This comment is conflicting with the one suggesting moving the functions out of LSP into their own space. If that change is made, this fake language server will not be necessary and should be removed instead as conflict solution.

Tobiky commented on tests/features/test_find_current_scope.py:
Please rename the tests according to what they're testing rather than numbering them, for example `test_find_current_scope_1` can instead be `test_find_scope_on_space_between_category_and_asset`, this makes it much easier when going through an overview to tell what's gone wrong.

Also move the file under unit since these are unit tests.

Tobiky reviewed (changes requested):
Couple of larger scale things that need addressing but other than that good implementation and documentation.

tagyieh committed:
chore: Move TreeSitter auxiliary functions to separate file.

tagyieh commented:
@Tobiky I believe these last couple of commits fix everything mentioned

tagyieh committed:
chore: Fixed tests.

Tobiky committed:
refactor: Move test/features files into respective integration or unit folders

Tobiky committed:
chore: Ruff check and format

Tobiky reviewed (approved):
Excellent job. Since it's so late into the night I added two commits to take care of some very small stuff for structure (move test files from the features folder into integration or unit tests folders) and refactoring (running ruff check and format)

Tobiky reviewed (changes requested):
Actually in the last minute I realized a thing that was missing. A text document position is not the same as a tree sitter point. They differe on the fact that LSP works on the [encoding based character position](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position) while [tree sitter works on pure bytes](https://github.com/tree-sitter/tree-sitter/issues/397). The function is supposed to take in the former as its the value that will be received on the server side. You can keep it the current way too if you also define a way to translate between the two. This should be relatively easy since its only the column that changes (though it does need to take into account newline settings) since row only cares about newline counts.

tagyieh committed:
feat: Conversion from LSP to TS positions.

tagyieh commented:
> Excellent job. Since it's so late into the night I added two commits to take care of some very small stuff for structure (move test files from the features folder into integration or unit tests folders) and refactoring (running ruff check and format)

I added a small helper function to convert from `UTF-16` to bytes which can be used when calling the scope finding function

tagyieh committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/find_current_scope' into feat/find_current_scope

Tobiky commented on src/malls/ts/utils.py:
```diff
@@ -92,3 +92,26 @@ def find_current_scope(cursor: TreeCursor, point: Point):
 
     return owner
+
+def lsp_to_tree_sitter(text: str, lsp_line: int, lsp_char: int) -> Point:
```
Use the `Position` from `models` instead of `lsp_line: int, lsp_char: int`

Tobiky commented on src/malls/ts/utils.py:
```diff
@@ -92,3 +92,26 @@ def find_current_scope(cursor: TreeCursor, point: Point):
+    line_text = lines[lsp_line]
+    
+    # Convert to UTF-16 little endian (each UTF-16 code unit is 2 bytes)
+    line_utf16 = line_text.encode('utf-16-le')
```
Why LE specifically? I couldn't find anything on the LSP requiring it and forcing LE on a non-LE platform is a big performance cost.

Tobiky commented on src/malls/ts/utils.py:
```diff
@@ -92,3 +92,26 @@ def find_current_scope(cursor: TreeCursor, point: Point):
+    string_slice = utf16_slice.decode('utf-16-le')
+    
+    # Encode the string slice to UTF-8 and get its byte length
+    byte_offset = len(string_slice.encode('utf-8'))
```
If `lsp_char * 2` gives the byte offset, why are further computations done? The additional computations would give the same results anyway on the basis of the comments. 

Furthermore, UTF-16 characters may be 1 or 2 16-bit values, so it may be x4 and not x2. Additionally, this will create an inconsistency for ASCII characters. ASCII is invariant to UTF-8 but not to UTF-16, that is to say ASCII is mostly a subset to UTF-8 but not at all to UTF-16. Some resources on the matter:
- [LSP on Text Documents](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments)
- [Wikipedia on UTF-16](https://en.wikipedia.org/wiki/UTF-16)

Since this function is very important for any feature interacting with text documents, please add some test to verify the correctness of it. They do not need to be large or many, as long as they test the core functionality of the UTF-16 to byte position conversion.

tagyieh commented:
@Tobiky since the conversion from LSP UTF-16 to TreeSitter bytes is of great importance, I think creating a separate issue, with its own commits and tests, might be worth it. Can I go ahead with it?

Tobiky commented:
> @Tobiky since the conversion from LSP UTF-16 to TreeSitter bytes is of great importance, I think creating a separate issue, with its own commits and tests, might be worth it. Can I go ahead with it?

@tagyieh Yeah sure, let's wait with this PR until that one is done then.

tagyieh commented:
@Tobiky Perfect. I am currently working on another issue, but once I am done with what I am doing I will dive right into it.

tagyieh committed:
Merge branch 'master' into feat/find_current_scope

tagyieh commented:
@Tobiky with #46 merged, is this PR also ready for merge, or does it need any correction?

tagyieh committed:
chore: Moved test to integration folder.

Tobiky commented:
> @Tobiky with #46 merged, is this PR also ready for merge, or does it need any correction?

I'll take a look at it shortly.

Tobiky commented on src/malls/ts/utils.py:
```diff
@@ -1,5 +1,98 @@
+    return compare_points(start_point, point)
+
+
+def find_current_scope(cursor: TreeCursor, point: Point):
```
Just need to change this to use a LSP location rather than a TS point, since the caller of this function will be a language server method.

tagyieh replied:
Considering that the server will receive the LSP Position and has the files/text saved, shouldn't the server use the auxiliary `lsp_to_tree_sitter_position` first (since it requires the file's text) and then pass the returned TS point to this `find_current_scope`?

With this I asking if this function should remain as is and move the responsibility of converting from a LSP to a TS position to the server, as it is the owner of the files.

Tobiky replied:
Sure that sounds reasonable. Do you mind jotting that down as a general principle too in the docs and then merge?

tagyieh committed:
docs: Explained position conversion and server's role.
